File: backend/services/groq_service.py
"""
Groq AI Service for resume analysis, interview questions, and candidate matching
"""

import os
import json
import re
from typing import Dict, List, Any, Optional
from groq import Groq
from models.schemas import (
    ResumeAnalysis, InterviewQuestions, InterviewQuestion, 
    InterviewAnalysis, CandidateMatch, SkillAlignment, Difficulty
)

class GroqService:
    """Service for interacting with Groq AI API"""
    
    def __init__(self):
        api_key = os.getenv("GROQ_API_KEY")
        if not api_key:
            raise ValueError("GROQ_API_KEY environment variable is required")
        
        self.client = Groq(api_key=api_key)
        self.model = "llama3-8b-8192"
    
    async def analyze_resume(self, resume_text: str, job_description: str = "") -> ResumeAnalysis:
        """Analyze resume text and provide comprehensive evaluation"""
        
        prompt = f"""
        Analyze the following resume and provide a comprehensive evaluation:
        
        Resume Text:
        {resume_text}
        
        {f"Job Description: {job_description}" if job_description else ""}
        
        Please provide:
        1. Skills extracted from the resume
        2. Years of experience estimate
        3. Key strengths and weaknesses
        4. Overall assessment score (1-100)
        5. Recommendations for improvement
        {f"6. Job fit analysis and compatibility score" if job_description else ""}
        
        Format your response as JSON with the following structure:
        {{
          "skills": ["skill1", "skill2", ...],
          "experience_years": number,
          "strengths": ["strength1", "strength2", ...],
          "weaknesses": ["weakness1", "weakness2", ...],
          "overall_score": number,
          "recommendations": ["rec1", "rec2", ...],
          {f'"job_fit_score": number,' if job_description else ""}
          "summary": "Brief summary of the candidate"
        }}
        """
        
        try:
            completion = self.client.chat.completions.create(
                messages=[{"role": "user", "content": prompt}],
                model=self.model,
                temperature=0.3,
                max_tokens=2000
            )
            
            response_content = completion.choices[0].message.content
            parsed_response = self._parse_json_response(response_content)
            
            return ResumeAnalysis(
                skills=parsed_response.get("skills", []),
                experience_years=parsed_response.get("experience_years", 0),
                strengths=parsed_response.get("strengths", []),
                weaknesses=parsed_response.get("weaknesses", []),
                overall_score=parsed_response.get("overall_score", 0),
                recommendations=parsed_response.get("recommendations", []),
                job_fit_score=parsed_response.get("job_fit_score"),
                summary=parsed_response.get("summary", "")
            )
            
        except Exception as e:
            print(f"Error analyzing resume: {e}")
            raise Exception("Failed to analyze resume")
    
    async def generate_interview_questions(
        self, 
        job_title: str, 
        job_description: str, 
        candidate_skills: List[str] = None, 
        difficulty: str = "medium"
    ) -> InterviewQuestions:
        """Generate interview questions for a position"""
        
        candidate_skills = candidate_skills or []
        
        prompt = f"""
        Generate interview questions for the following position:
        
        Job Title: {job_title}
        Job Description: {job_description}
        Candidate Skills: {', '.join(candidate_skills)}
        Difficulty Level: {difficulty}
        
        Please generate 10-15 interview questions covering:
        1. Technical skills (40%)
        2. Behavioral questions (30%)
        3. Situational questions (20%)
        4. Company culture fit (10%)
        
        Format your response as JSON:
        {{
          "technical": [
            {{"question": "...", "category": "technical", "difficulty": "easy|medium|hard"}}
          ],
          "behavioral": [
            {{"question": "...", "category": "behavioral", "difficulty": "easy|medium|hard"}}
          ],
          "situational": [
            {{"question": "...", "category": "situational", "difficulty": "easy|medium|hard"}}
          ],
          "cultural": [
            {{"question": "...", "category": "cultural", "difficulty": "easy|medium|hard"}}
          ]
        }}
        """
        
        try:
            completion = self.client.chat.completions.create(
                messages=[{"role": "user", "content": prompt}],
                model=self.model,
                temperature=0.7,
                max_tokens=3000
            )
            
            response_content = completion.choices[0].message.content
            parsed_response = self._parse_json_response(response_content)
            
            def parse_questions(questions_data: List[Dict]) -> List[InterviewQuestion]:
                return [
                    InterviewQuestion(
                        question=q.get("question", ""),
                        category=q.get("category", ""),
                        difficulty=Difficulty(q.get("difficulty", "medium"))
                    )
                    for q in questions_data
                ]
            
            return InterviewQuestions(
                technical=parse_questions(parsed_response.get("technical", [])),
                behavioral=parse_questions(parsed_response.get("behavioral", [])),
                situational=parse_questions(parsed_response.get("situational", [])),
                cultural=parse_questions(parsed_response.get("cultural", []))
            )
            
        except Exception as e:
            print(f"Error generating interview questions: {e}")
            raise Exception("Failed to generate interview questions")
    
    async def analyze_interview_response(
        self, 
        question: str, 
        answer: str, 
        job_context: str
    ) -> InterviewAnalysis:
        """Analyze an interview response"""
        
        prompt = f"""
        Analyze the following interview response:
        
        Question: {question}
        Answer: {answer}
        Job Context: {job_context}
        
        Please provide:
        1. Answer quality score (1-10)
        2. Key strengths in the response
        3. Areas for improvement
        4. Follow-up questions suggestions
        5. Overall assessment
        
        Format as JSON:
        {{
          "score": number,
          "strengths": ["strength1", "strength2", ...],
          "improvements": ["improvement1", "improvement2", ...],
          "follow_up_questions": ["question1", "question2", ...],
          "assessment": "Overall assessment text"
        }}
        """
        
        try:
            completion = self.client.chat.completions.create(
                messages=[{"role": "user", "content": prompt}],
                model=self.model,
                temperature=0.3,
                max_tokens=1500
            )
            
            response_content = completion.choices[0].message.content
            parsed_response = self._parse_json_response(response_content)
            
            return InterviewAnalysis(
                score=parsed_response.get("score", 0),
                strengths=parsed_response.get("strengths", []),
                improvements=parsed_response.get("improvements", []),
                follow_up_questions=parsed_response.get("follow_up_questions", []),
                assessment=parsed_response.get("assessment", "")
            )
            
        except Exception as e:
            print(f"Error analyzing interview response: {e}")
            raise Exception("Failed to analyze interview response")
    
    async def match_candidate_to_job(
        self, 
        candidate_profile: Dict[str, Any], 
        job_description: str
    ) -> CandidateMatch:
        """Analyze compatibility between candidate and job position"""
        
        prompt = f"""
        Analyze the compatibility between this candidate and job position:
        
        Candidate Profile:
        {json.dumps(candidate_profile, indent=2)}
        
        Job Description:
        {job_description}
        
        Please provide:
        1. Overall match score (1-100)
        2. Skill alignment analysis
        3. Experience level fit
        4. Potential red flags or concerns
        5. Strengths that make them a good fit
        6. Areas where they might need development
        
        Format as JSON:
        {{
          "match_score": number,
          "skill_alignment": {{
            "matched": ["skill1", "skill2", ...],
            "missing": ["skill1", "skill2", ...],
            "bonus": ["skill1", "skill2", ...]
          }},
          "experience_fit": "underqualified|qualified|overqualified",
          "red_flags": ["flag1", "flag2", ...],
          "strengths": ["strength1", "strength2", ...],
          "development_areas": ["area1", "area2", ...],
          "recommendation": "hire|interview|reject",
          "reasoning": "Detailed reasoning for the recommendation"
        }}
        """
        
        try:
            completion = self.client.chat.completions.create(
                messages=[{"role": "user", "content": prompt}],
                model=self.model,
                temperature=0.3,
                max_tokens=2000
            )
            
            response_content = completion.choices[0].message.content
            parsed_response = self._parse_json_response(response_content)
            
            skill_alignment_data = parsed_response.get("skill_alignment", {})
            skill_alignment = SkillAlignment(
                matched=skill_alignment_data.get("matched", []),
                missing=skill_alignment_data.get("missing", []),
                bonus=skill_alignment_data.get("bonus", [])
            )
            
            return CandidateMatch(
                match_score=parsed_response.get("match_score", 0),
                skill_alignment=skill_alignment,
                experience_fit=parsed_response.get("experience_fit", "qualified"),
                red_flags=parsed_response.get("red_flags", []),
                strengths=parsed_response.get("strengths", []),
                development_areas=parsed_response.get("development_areas", []),
                recommendation=parsed_response.get("recommendation", "interview"),
                reasoning=parsed_response.get("reasoning", "")
            )
            
        except Exception as e:
            print(f"Error matching candidate to job: {e}")
            raise Exception("Failed to match candidate to job")
    
    def _parse_json_response(self, response: str) -> Dict[str, Any]:
        """Parse JSON response from AI, handling various formats"""
        try:
            # Try to extract JSON from the response
            json_match = re.search(r'\{.*\}', response, re.DOTALL)
            if json_match:
                json_str = json_match.group(0)
                return json.loads(json_str)
            
            # If no JSON found, return error response
            return {
                "error": "Invalid response format",
                "raw_response": response
            }
            
        except json.JSONDecodeError as e:
            print(f"Error parsing JSON response: {e}")
            return {
                "error": "Failed to parse AI response",
                "raw_response": response
            }

# Create a singleton instance
groq_service = GroqService()

File: server/services/groqService.js
const Groq = require('groq-sdk');

class GroqService {
  constructor() {
    this.client = new Groq({
      apiKey: process.env.GROQ_API_KEY
    });
  }

  async analyzeResume(resumeText, jobDescription = '') {
    try {
      const prompt = `
        Analyze the following resume and provide a comprehensive evaluation. Pay special attention to extracting work experience accurately:
        
        Resume Text:
        ${resumeText}
        
        ${jobDescription ? `Job Description: ${jobDescription}` : ''}
        
        For experience calculation, consider ALL formats:
        - "2+ years experience" = 2 years
        - "5-7 years" = 6 years (average)
        - "2023-Present" = calculate from 2023 to current year
        - "Jan 2022 - Dec 2023" = calculate the difference
        - "Recent graduate" = 0 years
        - Multiple jobs = sum total experience
        - Overlapping positions = don't double count
        
        Please provide:
        1. Skills extracted from the resume (programming languages, frameworks, tools, soft skills)
        2. Total years of experience (calculate accurately from all work history)
        3. Current/most recent job title and company
        4. Education level and field
        5. Key strengths and weaknesses
        6. Overall assessment score (1-100)
        7. Recommendations for improvement
        ${jobDescription ? '8. Job fit analysis and compatibility score' : ''}
        
        Format your response as JSON with the following structure:
        {
          "skills": ["skill1", "skill2", ...],
          "experienceYears": number,
          "currentJobTitle": "string",
          "currentCompany": "string", 
          "educationLevel": "string",
          "educationField": "string",
          "strengths": ["strength1", "strength2", ...],
          "weaknesses": ["weakness1", "weakness2", ...],
          "overallScore": number,
          "recommendations": ["rec1", "rec2", ...],
          ${jobDescription ? '"jobFitScore": number,' : ''}
          "summary": "Brief summary of the candidate",
          "experienceBreakdown": "Detailed explanation of how experience was calculated"
        }
      `;

      const completion = await this.client.chat.completions.create({
        messages: [{ role: 'user', content: prompt }],
        model: 'llama-3.3-70b-versatile',
        temperature: 0.3,
        max_tokens: 2000
      });

      const response = completion.choices[0]?.message?.content;
      return this.parseJsonResponse(response);
    } catch (error) {
      console.error('Error analyzing resume:', error);
      throw new Error('Failed to analyze resume');
    }
  }

  async generateInterviewQuestions(jobTitle, jobDescription, candidateSkills = [], difficulty = 'easy') {
    try {
      const prompt = `
You are creating interview questions for a VOICE-RECORDED interview session. The output MUST be valid JSON and nothing else.

Important override: Regardless of the jobDifficulty input, set every question's "difficulty" field to "easy". All questions should be voice-friendly and suitable for short spoken answers.

Context:
- This interview will be read aloud by an interviewer and recorded as voice responses from the candidate.
- Questions must be short, clear, and natural to speak (suitable for TTS or an interviewer reading them).
- Candidate answers will be short-to-medium. Provide guidance for expected answer length (seconds).
- Include a short "readable" prompt (1 sentence) the interviewer can read exactly.
- Include a recommended pause (in milliseconds) the interviewer should wait after reading the question to allow the candidate to respond.

Requirements:
1. Generate between 10 and 15 questions total.
2. Distribute questions roughly as: Technical 40%, Behavioral 30%, Situational 20%, Cultural 10%.
3. For each question include:
   - "question": the full question text (suitable for internal display)
   - "readable": a concise single-sentence phrasing the interviewer should read aloud (≤ 20 words)
   - "category": one of "technical", "behavioral", "situational", "cultural"
   - "difficulty": **must be** "easy" for every question
   - "expected_answer_duration_seconds": integer (recommended seconds candidate should speak; typical easy ranges 15-60)
   - "recommended_pause_ms": integer (milliseconds to pause after reading; for easy questions recommend 8000-30000)
   - "follow_up": an array of up to 2 short follow-up question strings (optional)
4. Keep each "readable" phrase free of special characters that might confuse TTS.
5. Make questions appropriate for the provided inputs:
   - Job Title: ${jobTitle}
   - Job Description: ${jobDescription}
   - Candidate Skills: ${candidateSkills.join(', ')}
   - Overall difficulty preference: ${difficulty} (IGNORE this for per-question difficulty; still include in meta)
6. Ensure technical questions test core skills listed in Candidate Skills and the Job Description. Prioritize practical, conversational prompts that can be answered aloud.
7. Avoid numbering inside the question strings (the JSON structure will indicate order).
8. Output only JSON structured exactly as below.

Desired JSON format:
{
  "meta": {
    "jobTitle": "...",
    "difficulty": "easy|medium|hard", // keep the original field but per-question fields must be "easy"
    "total_questions": 12,
    "notes": "Voice-recording friendly; all questions difficulty='easy'; 'readable' is for interviewer."
  },
  "technical": [ ... ],
  "behavioral": [ ... ],
  "situational": [ ... ],
  "cultural": [ ... ]
}

Final instructions for the model:
- Do not include any explanatory text outside the JSON.
- Ensure JSON is valid and parseable.
- Set every question's "difficulty" property to the string "easy".
- Make expected_answer_duration_seconds and recommended_pause_ms realistic for spoken easy questions.
- Keep question wording voice-friendly and concise.
`;

      const completion = await this.client.chat.completions.create({
        messages: [{ role: 'user', content: prompt }],
        model: 'llama-3.3-70b-versatile',
        temperature: 0.7,
        max_tokens: 3000
      });

      const response = completion.choices[0]?.message?.content;
      return this.parseJsonResponse(response);
    } catch (error) {
      console.error('Error generating interview questions:', error);
      throw new Error('Failed to generate interview questions');
    }
  }

  async analyzeAssessmentResponses(questions, responses) {
    try {
      const prompt = `You are an expert HR analyst. Analyze the following interview assessment responses and provide detailed feedback.
  
  QUESTIONS AND RESPONSES:
  ${JSON.stringify({ questions, responses }, null, 2)}
  
  Please provide a comprehensive analysis in the following JSON format:
  {
    "overallScore": <number 0-100>,
    "overallFeedback": "<detailed overall assessment>",
    "sectionAnalysis": {
      "technical": {
        "score": <number 0-100>,
        "feedback": "<detailed feedback>",
        "strengths": ["<strength 1>", "<strength 2>"],
        "weaknesses": ["<weakness 1>", "<weakness 2>"],
        "responses": [
          {
            "question": "<question text>",
            "answer": "<answer text>",
            "score": <number 0-10>,
            "feedback": "<specific feedback for this answer>"
          }
        ]
      },
      "behavioral": {
        "score": <number 0-100>,
        "feedback": "<detailed feedback>",
        "strengths": ["<strength 1>", "<strength 2>"],
        "weaknesses": ["<weakness 1>", "<weakness 2>"],
        "responses": [
          {
            "question": "<question text>",
            "answer": "<answer text>",
            "score": <number 0-10>,
            "feedback": "<specific feedback>"
          }
        ]
      },
      "situational": {
        "score": <number 0-100>,
        "feedback": "<detailed feedback>",
        "strengths": ["<strength 1>", "<strength 2>"],
        "weaknesses": ["<weakness 1>", "<weakness 2>"],
        "responses": [
          {
            "question": "<question text>",
            "answer": "<answer text>",
            "score": <number 0-10>,
            "feedback": "<specific feedback>"
          }
        ]
      },
      "cultural": {
        "score": <number 0-100>,
        "feedback": "<detailed feedback>",
        "strengths": ["<strength 1>", "<strength 2>"],
        "weaknesses": ["<weakness 1>", "<weakness 2>"],
        "responses": [
          {
            "question": "<question text>",
            "answer": "<answer text>",
            "score": <number 0-10>,
            "feedback": "<specific feedback>"
          }
        ]
      }
    },
    "recommendation": "<Schedule an Interview|REJECT>",
    "recommendationReason": "<detailed explanation of recommendation>",
    "keyTakeaways": ["<takeaway 1>", "<takeaway 2>", "<takeaway 3>"]
  }
  
  Evaluate each response based on:
  - Relevance and completeness
  - Technical accuracy (for technical questions)
  - Communication clarity
  - Problem-solving approach
  - Cultural fit alignment
  
  Provide constructive feedback and actionable insights.`;
  
      const completion = await this.client.chat.completions.create({
        messages: [
          {
            role: 'system',
            content: 'You are an expert HR analyst specializing in interview assessment and candidate evaluation. Provide detailed, constructive feedback in valid JSON format only.'
          },
          {
            role: 'user',
            content: prompt
          }
        ],
        model: 'llama-3.3-70b-versatile',
        temperature: 0.3,
        max_tokens: 8000,
        response_format: { type: 'json_object' }
      });
  
      const analysisText = completion.choices[0]?.message?.content;
      if (!analysisText) {
        throw new Error('No analysis generated');
      }

      const analysis = JSON.parse(analysisText);
      
      return analysis;
    } catch (error) {
      console.error('Assessment analysis error:', error);
      throw new Error(`Failed to analyze assessment: ${error.message}`);
    }
  }

  async analyzeInterviewResponse(question, answer, jobContext) {
    try {
      const prompt = `
        Analyze the following interview response:
        
        Question: ${question}
        Answer: ${answer}
        Job Context: ${jobContext}
        
        Please provide:
        1. Answer quality score (1-10)
        2. Key strengths in the response
        3. Areas for improvement
        4. Follow-up questions suggestions
        5. Overall assessment
        
        Format as JSON:
        {
          "score": number,
          "strengths": ["strength1", "strength2", ...],
          "improvements": ["improvement1", "improvement2", ...],
          "followUpQuestions": ["question1", "question2", ...],
          "assessment": "Overall assessment text"
        }
      `;

      const completion = await this.client.chat.completions.create({
        messages: [{ role: 'user', content: prompt }],
        model: 'llama-3.3-70b-versatile',
        temperature: 0.3,
        max_tokens: 1500
      });

      const response = completion.choices[0]?.message?.content;
      return this.parseJsonResponse(response);
    } catch (error) {
      console.error('Error analyzing interview response:', error);
      throw new Error('Failed to analyze interview response');
    }
  }

  async matchCandidateToJob(candidateProfile, jobDescription) {
    try {
      const prompt = `
        Analyze the compatibility between this candidate and job position:
        
        Candidate Profile:
        ${JSON.stringify(candidateProfile, null, 2)}
        
        Job Description:
        ${jobDescription}
        
        Please provide:
        1. Overall match score (1-100)
        2. Skill alignment analysis
        3. Experience level fit
        4. Potential red flags or concerns
        5. Strengths that make them a good fit
        6. Areas where they might need development
        
        Format as JSON:
        {
          "matchScore": number,
          "skillAlignment": {
            "matched": ["skill1", "skill2", ...],
            "missing": ["skill1", "skill2", ...],
            "bonus": ["skill1", "skill2", ...]
          },
          "experienceFit": "underqualified|qualified|overqualified",
          "redFlags": ["flag1", "flag2", ...],
          "strengths": ["strength1", "strength2", ...],
          "developmentAreas": ["area1", "area2", ...],
          "recommendation": "hire|interview|reject",
          "reasoning": "Detailed reasoning for the recommendation"
        }
      `;

      const completion = await this.client.chat.completions.create({
        messages: [{ role: 'user', content: prompt }],
        model: 'llama-3.3-70b-versatile',
        temperature: 0.3,
        max_tokens: 2000
      });

      const response = completion.choices[0]?.message?.content;
      return this.parseJsonResponse(response);
    } catch (error) {
      console.error('Error matching candidate to job:', error);
      throw new Error('Failed to match candidate to job');
    }
  }

  async autoMatchCandidateToJobs(candidateProfile, availableJobs) {
    try {
      const prompt = `
        Analyze this candidate profile and automatically match them to the most suitable job positions:
        
        Candidate Profile:
        ${JSON.stringify(candidateProfile, null, 2)}
        
        Available Job Positions:
        ${JSON.stringify(availableJobs, null, 2)}
        
        For each job, calculate:
        1. Match score (0-100) based on:
           - Skills alignment (40%)
           - Experience level fit (30%)
           - Education requirements (20%)
           - Other factors (10%)
        2. Detailed reasoning for the score
        3. Missing skills that candidate should develop
        4. Recommendation (hire/interview/reject)
        
        Only return matches with score >= 30. Sort by match score descending.
        
        Format as JSON:
        {
          "matches": [
            {
              "jobId": "string",
              "jobTitle": "string",
              "matchScore": number,
              "skillsMatched": ["skill1", "skill2"],
              "skillsMissing": ["skill1", "skill2"],
              "experienceFit": "underqualified|qualified|overqualified",
              "recommendation": "hire|interview|reject",
              "reasoning": "Detailed explanation of the match"
            }
          ],
          "summary": "Overall assessment of candidate's job market fit"
        }
      `;

      const completion = await this.client.chat.completions.create({
        messages: [{ role: 'user', content: prompt }],
        model: 'llama-3.3-70b-versatile',
        temperature: 0.3,
        max_tokens: 3000
      });

      const response = completion.choices[0]?.message?.content;
      return this.parseJsonResponse(response);
    } catch (error) {
      console.error('Error in auto-matching:', error);
      throw new Error('Failed to auto-match candidate to jobs');
    }
  }

  parseJsonResponse(response) {
    try {
      // Try to extract JSON from the response
      const jsonMatch = response.match(/\{[\s\S]*\}/);
      if (jsonMatch) {
        return JSON.parse(jsonMatch[0]);
      }
      
      // If no JSON found, return a structured error response
      return {
        error: 'Invalid response format',
        rawResponse: response
      };
    } catch (error) {
      console.error('Error parsing JSON response:', error);
      return {
        error: 'Failed to parse AI response',
        rawResponse: response
      };
    }
  }
}

module.exports = new GroqService();